src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+	 * ON CONFLICT/RETURNING clauses. If the query is of such a query we need to consider
+	 * pruning intermediate results even though we do not have a workerjob field
+	 *
+	 *  * We do not wish to proceed if one of the following is met:
+	 * - there are no intermediate results
+	 * - the query hits all the worker nodes
+	 */
+	if (jobQuery == NULL)
+	{
+		return;
+	}
+
+	rangeTableList = ExtractRangeTableEntryList(jobQuery);
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = lfirst(rangeTableCell);
+		if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			Const *resultIdConst = NULL;
+			Datum resultIdDatum = 0;
+			char *resultId = NULL;
+
+			resultIdConst = FindResultIdOfIntermediateResultFunction(rangeTableEntry);
+
+			if (resultIdConst == NULL)
+			{
+				continue;
+			}
+
+			plan->usedSubPlanNodeList = list_append_unique(plan->usedSubPlanNodeList,
+														   resultIdConst);
+
+			resultIdDatum = resultIdConst->constvalue;
+			resultId = TextDatumGetCString(resultIdDatum);
+			elog(DEBUG4, "appending %s subplan to usedSubPlanNodeList of plan %lu",
+				 resultId, plan->planId);
+		}
+	}
+}
+
+
+/*
+ * RecordSubplanExecutionsOnNodes iterates over the usedSubPlanNodeList,
+ * and for each entry, record the workerNodes that are accessed by
+ * the distributed plan.
+ *
+ * Later, we'll use this information while we broadcast the intermediate
+ * results to the worker nodes. The idea is that the intermediate result
+ * should only be broadcasted to the worker nodes that are accessed by
+ * the distributedPlan(s) that the subPlan is used in.
+ *
+ * Finally, the function recursively descends into the actual subplans
+ * of the input distributedPlan as well.
+ */
+void
+RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
+							   DistributedPlan *distributedPlan)
+{
+	List *usedSubPlanNodeList = distributedPlan->usedSubPlanNodeList;
+	ListCell *usedSubPlanNodeCell = NULL;
+	List *subPlanList = distributedPlan->subPlanList;
+	ListCell *subPlanCell = NULL;
+	int workerNodeCount = GetWorkerNodeCount();
+
+	foreach(usedSubPlanNodeCell, usedSubPlanNodeList)
+	{
+		Const *resultIdConst = lfirst(usedSubPlanNodeCell);
+		Datum resultIdDatum = resultIdConst->constvalue;
+		char *resultId = TextDatumGetCString(resultIdDatum);
+
+		IntermediateResultsHashEntry *entry = SearchIntermediateResult(
+			intermediateResultsHash, resultId);
+
+		/* no need to traverse the whole plan if all the workers are hit */
+		if (list_length(entry->nodeIdList) == workerNodeCount)
+		{
+			elog(DEBUG4, "subplan %s is used in all workers", resultId);
+
+			break;
+		}
+		else
+		{
+			entry->nodeIdList = AppendAllAccessedWorkerNodes(entry->nodeIdList,
+															 distributedPlan,
+															 workerNodeCount);
+
+			elog(DEBUG4, "subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
```

Comment:
is this still necessary if we send to all nodes?

Reply:
`RecordSubplanExecutionsOnNodes()` is recursively called for every subplan of a plan. Currently at every call we check node count with `GetWorkerNodeCount()`

I can easily add a new parameter to this recursive function to keep the number of nodes, and possibly create a wrapper that can initialize the node count and pass it to the first call of the original function.

Do you think increasing the complexity here is worth it?